Sources/PIRProcessDatabase/ProcessDatabase.swift

```diff
@@ -411,14 +386,78 @@ struct ProcessDatabase: ParsableCommand {
         }
     }
 
-    mutating func run() throws {
+    private func processShard<Scheme: HeScheme>(
+        shardID: String,
+        shard: KeywordDatabaseShard,
+        config: ResolvedArguments,
+        context: Context<Scheme>,
+        processArgs: ProcessKeywordDatabase.Arguments<Scheme>) async throws -> EvaluationKeyConfig
+    {
+        func logEvent(event: ProcessKeywordDatabase.ProcessShardEvent) throws {
+            let log: (String) -> Void = { message in ProcessDatabase.logger.info("Shard №\(shardID): \(message)") }
+            switch event {
+            case let .cuckooTableEvent(.createdTable(table)):
+                let summary = try table.summarize()
+                log("Created cuckoo table \(summary)")
+            case let .cuckooTableEvent(.expandingTable(table)):
+                let summary = try table.summarize()
+                log("Expanding cuckoo table \(summary)")
+            case let .cuckooTableEvent(.finishedExpandingTable(table)):
+                let summary = try table.summarize()
+                log("Finished expanding cuckoo table \(summary)")
+            case let .cuckooTableEvent(.insertedKeywordValuePair(index, _)):
+                let reportingPercentage = 10
+                let shardFraction = shard.rows.count / reportingPercentage
+                if (index + 1).isMultiple(of: shardFraction) {
+                    let percentage = Float(reportingPercentage * (index + 1)) / Float(shardFraction)
+                    log("Inserted \(index + 1) / \(shard.rows.count) keywords \(percentage)%")
+                }
+            }
+        }
+
+        ProcessDatabase.logger.info("Processing shard \(shardID) with \(shard.rows.count) rows")
```

Comment:
This could also use the local instance of logger that adds in `shardID` to the metadata.

Even further, every use of `ProcessDatabase.logger` in this function should be replaced with a call to local logger that has `shardID` in the metadata.

Reply:
I’ve made the changes using the local logger with metadata. Just to clarify—did you mean to remove all the concatenations of `\(shardID)` as well? If needed, I can update that too, let me know.

Reply:
Yes, remove them. We do not need duplicated information in the same log message.

Reply:
Done ✅